### ISSUE TITLE
A few questions for me to learn

### ISSUE THREAD

**Issue by Optimox** [haqishen/SIIM-ISIC-Melanoma-Classification-1st-Place-Solution#1]
Thank you very much for sharing the 1st place solution so quickly, congratulation for winning the competition and for this nice code sharing.

I just wanted to ask a few questions that I have when reading your code, I hope you'll find some time to answer them.

-  I see here https://github.com/haqishen/SIIM-ISIC-Melanoma-Classification-1st-Place-Solution/blob/master/train.py#L84-L86 that you only perform gradient clipping for some image sizes, would you mind explaining why those two sizes (896 - 576)? Is this because they are bigger? Didn't you train with 768x768 images? Also in general how do you choose the clipping value?

-  You are using Swish activation https://github.com/haqishen/SIIM-ISIC-Melanoma-Classification-1st-Place-Solution/blob/master/models.py#L11-L26, would you say that in general Swish is always a better choice that basic ReLU?

- I'd like to understand what you are doing here : https://github.com/haqishen/SIIM-ISIC-Melanoma-Classification-1st-Place-Solution/blob/master/models.py#L61-L66 It seems that you are applying 5 different dropouts at the very end on the same linear layer and average them, it does not seem a very common approach. I'm seeing this as a way of making sure that the last layer is training correctly but is this really needed? I find interesting the idea of self ensembling a model with different random heads at the end but why do you use the same linear layer? Why not having 5 different heads with different random inputs during training? I'd really like to understand :) Have you ever tried the 5 linear layer approach?

Thanks in advance!

Cheers

**Comment by datduong** [comment]
I also want to ask, why do not need to use weights in nn.CrossEntropyLoss to account for imbalance numbers of labels.

**Comment by haqishen** [comment]
Sorry for the late reply since I didn't check my github so often and here's my answer.

> I see here https://github.com/haqishen/SIIM-ISIC-Melanoma-Classification-1st-Place-Solution/blob/master/train.py#L84-L86 that you only perform gradient clipping for some image sizes, would you mind explaining why those two sizes (896 - 576)? Is this because they are bigger? Didn't you train with 768x768 images? Also in general how do you choose the clipping value?

Those specific model and image size raised some bug while training, which is caused by apex amp (but didn't know why). gradient clip helps to prevent it.

> You are using Swish activation https://github.com/haqishen/SIIM-ISIC-Melanoma-Classification-1st-Place-Solution/blob/master/models.py#L11-L26, would you say that in general Swish is always a better choice that basic ReLU?

To my limited experience, yes.

> I'd like to understand what you are doing here : https://github.com/haqishen/SIIM-ISIC-Melanoma-Classification-1st-Place-Solution/blob/master/models.py#L61-L66 It seems that you are applying 5 different dropouts at the very end on the same linear layer and average them, it does not seem a very common approach. I'm seeing this as a way of making sure that the last layer is training correctly but is this really needed? I find interesting the idea of self ensembling a model with different random heads at the end but why do you use the same linear layer? Why not having 5 different heads with different random inputs during training? I'd really like to understand :) Have you ever tried the 5 linear layer approach?

Good idea! We didnt' try it before. Maybe will try it next time.

> why do not need to use weights in nn.CrossEntropyLoss to account for imbalance numbers of labels.

Since we only use one probability out of 9 classes and we use auc to evaluate the performance. So we dont' need to consider the imbalance.

**Comment by Optimox** [comment]
thank you @haqishen!
